### PR TITLE
User controller spec (part 2)

### PR DESCRIPTION
**Pull request by mdeniz**
More specs for user_controller_test. Needed to create and modify some factories.

### CODE REVIEW

**Comment by ChrisBr** [comment]
Do we really need all the cassettes? 24k line adds ... Uff

**Comment by mdeniz** [comment]
If you don't want the backend running everytime... yes

**Comment by hennevogel** [comment]
None of the requests in the cassettes go to the backend. They all go to gravatar to fetch user icons. You should ignore this type of request like we do already for others in [spec/support/vcr](https://github.com/openSUSE/open-build-service/blob/master/src/api/spec/support/vcr.rb) or  you mock them with [webmock](https://github.com/bblimke/webmock) directly.

**Comment by mdeniz** [comment]
@hennevogel  Done, VCR ignores gravatar.com now

File: src/api/spec/features/webui/users/users_ui_spec.rb [hunk 1]
```diff
@@ -0,0 +1,62 @@
+require "browser_helper"
+
+RSpec.feature "User's UI shows", type: :feature, js: true do
```
Comment (hennevogel) [inline src/api/spec/features/webui/users/users_ui_spec.rb]: `js: true` is default for feature specs

Comment (mdeniz) [inline src/api/spec/features/webui/users/users_ui_spec.rb]: Ok, then we should change it in almost all the feature tests...

File: src/api/spec/features/webui/users/users_ui_spec.rb [hunk 2]
```diff
@@ -0,0 +1,62 @@
+require "browser_helper"
```
Comment (hennevogel) [inline src/api/spec/features/webui/users/users_ui_spec.rb]: I must say I dislike the naming. users_ui. User dashboard maybe?

File: src/api/spec/factories/attrib_types.rb [hunk 3]
```diff
@@ -5,5 +5,10 @@
     factory :attrib_type_with_namespace do
       attrib_namespace { create(:attrib_namespace) }
     end
+
+    factory :attrib_type_OBS_OwnerRootProject do
```
Comment (hennevogel) [inline src/api/spec/factories/attrib_types.rb]: Why not name it like the attribute? OBS:OwnerRootProject?

File: src/api/spec/features/webui/users/users_ui_spec.rb (hunk 1 above)
Comment (mdeniz) [inline src/api/spec/features/webui/users/users_ui_spec.rb]: And if I change it ... it fails.

File: src/api/spec/features/webui/users/users_notifications_settings_spec.rb [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+    expect(page).to have_content 'Notifications settings updated'
+    expect(page).to have_content 'Get mails if in group'
+    expect(page).to have_unchecked_field(group_title)
+  end
+
+  scenario "when a user isn't in any group" do
+    login user
+    visit user_notifications_path
+
+    expect(page).not_to have_content 'Get mails if in group'
```
Comment (hennevogel) [inline src/api/spec/features/webui/users/users_notifications_settings_spec.rb]: What exactly are you testing here? The .unless in the view?

File: src/api/spec/factories/attrib_types.rb (hunk 3 above)
Comment (mdeniz) [inline src/api/spec/factories/attrib_types.rb]: Because is an attribute type... ;)

File: src/api/spec/features/webui/users/users_ui_spec.rb (hunk 2 above)
Comment (mdeniz) [inline src/api/spec/features/webui/users/users_ui_spec.rb]: Ok I will change it

File: src/api/spec/features/webui/users/users_notifications_settings_spec.rb [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+    check 'Event::CommentForProject_commenter'
+    click_button 'Update'
+
+    expect(page).to have_content 'Notifications settings updated'
+    expect(page).to have_content 'Events to get email for'
+    expect(page).to have_checked_field 'Event::RequestStatechange_creator'
+    expect(page).to have_checked_field 'Event::CommentForPackage_maintainer'
+    expect(page).to have_checked_field 'Event::CommentForPackage_commenter'
+    expect(page).to have_checked_field 'Event::CommentForProject_maintainer'
+    expect(page).to have_checked_field 'Event::CommentForProject_commenter'
```
Comment (hennevogel) [inline src/api/spec/features/webui/users/users_notifications_settings_spec.rb]: Isn't this better tested in a controller test?

File: src/api/spec/features/webui/users/users_ui_spec.rb [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+  let(:user) { create(:confirmed_user, login: 'moi') }
+  let(:another_user) { create(:confirmed_user, login: 'henne') }
+  let(:project) { create(:project, name: 'project_a') }
+  let(:project_with_package) { create(:project_with_package, name: 'project_b') }
+
+  before do
+    login user
+  end
+
+  scenario "a table with involved packages" do
```
Comment (hennevogel) [inline src/api/spec/features/webui/users/users_ui_spec.rb]: I think you should test User.involved_packages instead

File: src/api/spec/features/webui/users/users_ui_spec.rb [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+    create(:relationship_package_user, package: project_with_package.packages.first, user: another_user)
+    visit user_show_path(user: another_user)
+
+    within "table#ipackages_wrapper_table" do
+      expect(find(:xpath, './/tr[1]/td[1]').text).to eq(project_with_package.packages.first.name)
+      expect(find(:xpath, './/tr[1]/td[2]').text).to eq(project_with_package.name)
+    end
+  end
+
+  scenario "a table with involved projects" do
```
Comment (hennevogel) [inline src/api/spec/features/webui/users/users_ui_spec.rb]: I think you should test User#involved_projects instead

File: src/api/spec/features/webui/users/users_ui_spec.rb [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+    within "table#projects_table" do
+      expect(find(:xpath, './/tr[1]/td[1]').text).to eq(another_user.home_project_name)
+      expect(find(:xpath, './/tr[2]/td[1]').text).to eq(project.name)
+      expect(find(:xpath, './/tr[2]/td[2]').text).to eq(project.title)
+      expect(find(:xpath, './/tr[3]/td[1]').text).to eq(project_with_package.name)
+      expect(find(:xpath, './/tr[3]/td[2]').text).to eq(project_with_package.title)
+    end
+  end
+
+  scenario "a table with owned projects and packages" do
```
Comment (hennevogel) [inline src/api/spec/features/webui/users/users_ui_spec.rb]: I think you should test User#owned_packages instead

File: src/api/spec/features/webui/users/users_ui_spec.rb [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+    visit user_show_path(user: another_user)
+
+    click_link("Owned Project/Packages")
+    within "table#iowned_wrapper_table" do
+      expect(find(:xpath, './/tr[1]/td[2]').text).to eq(project_with_package.name)
+      expect(find(:xpath, './/tr[2]/td[1]').text).to eq(project_with_package.packages.first.name)
+      expect(find(:xpath, './/tr[2]/td[2]').text).to eq(project_with_package.name)
+    end
+  end
+
```
Comment (hennevogel) [inline src/api/spec/features/webui/users/users_ui_spec.rb]: And if you test the code in a Model spec you can test the different conditions that exist in the view in this feature spec.

**Comment by hennevogel** [comment]
@mdeniz you still include the cassettes which are not needed anymore if you ignore the request :-)

**Comment by mdeniz** [comment]
I don't see those unneeded cassettes included... I remove them and rerun the whole thing with the backend running.. so the cassettes there are what should be..aren't them?

File: src/api/spec/features/webui/users/users_ui_spec.rb (hunk 1 above)
Comment (hennevogel) [inline src/api/spec/features/webui/users/users_ui_spec.rb]: Oh shit I was wrong! It was [`:vcr`](https://github.com/openSUSE/open-build-service/blob/master/src/api/spec/spec_helper.rb#L65) that is default not `:js`.  Sorry for the noise

**Comment by hennevogel** [comment]
Ah true those are different ones. I wouldn't do that as this changes cassettes like ` spec/cassettes/Sign_up/User.yml ` which are order dependent and changes only the timestamp in others.

File: src/api/spec/features/webui/users/users_notifications_settings_spec.rb (hunk 4 above)
Comment (mdeniz) [inline src/api/spec/features/webui/users/users_notifications_settings_spec.rb]: @hennevogel Just translated it to RSpec...
But, is a good question.. Should we test the opposite cases?

Comment (hennevogel) [inline src/api/spec/features/webui/users/users_notifications_settings_spec.rb]: Don't just copy things over. Please check what makes sense and what doesn't, and if test make sense at the place where they are.

In this case I would say that testing a simple `unless` statement in a view is pointless...

File: src/api/spec/features/webui/users/users_notifications_settings_spec.rb (hunk 5 above)
Comment (mdeniz) [inline src/api/spec/features/webui/users/users_notifications_settings_spec.rb]: In this case is already covered by the User model spec done by Christian... So I will remove this

**Comment by hennevogel** [comment]
Please do not drop the`Sign_up/User.yml` cassette

**Comment by mdeniz** [comment]
Now is there @hennevogel

File: src/api/spec/factories/groups.rb [hunk 10]
```diff
@@ -0,0 +1,6 @@
+FactoryGirl.define do
+  factory :group do
+    title { Faker::Internet.user_name(nil, %w(. -))  }
```
Comment (ChrisBr) [inline src/api/spec/factories/groups.rb]: title must be unique, so I'd recommend to use a sequence here:

```
sequence(:title){|n| "#{Faker::Internet.user_name(nil, %w(_))}#{n}" }
```

File: src/api/spec/features/webui/users/users_notifications_settings_spec.rb [hunk 11]
```diff
@@ -0,0 +1,19 @@
+require "browser_helper"
+
+RSpec.feature "User's notifications settings", type: :feature, js: true do
+  let!(:user_with_groups) { create(:user_with_groups, login: 'moi_wg') }
```
Comment (ChrisBr) [inline src/api/spec/features/webui/users/users_notifications_settings_spec.rb]: why ``let!`` ? Isn't ```let()`` sufficient?

**Comment by ChrisBr** [comment]
otherwise LGTM

File: src/api/spec/factories/groups.rb (hunk 10 above)
Comment (mdeniz) [inline src/api/spec/factories/groups.rb]: Ok

File: src/api/spec/features/webui/users/users_notifications_settings_spec.rb (hunk 11 above)
Comment (mdeniz) [inline src/api/spec/features/webui/users/users_notifications_settings_spec.rb]: Now make sense... before was used in other specs inside this file (the were dropped)

**Comment by hennevogel** [comment]
:fire: frei!

**Pull request merged**